src/ultk/effcomm/information_bottleneck/ib.py

```diff
@@ -0,0 +1,181 @@
+"""Information Bottleneck optimizer."""
+
+import numpy as np
+
+from collections import namedtuple
+from scipy.special import logsumexp, log_softmax
+
+from .ba import BaseRDOptimizer
+from .tools import PRECISION, information_cond, mutual_info, kl_divergence
+
+
+def ib_kl(py_x: np.ndarray, qy_xhat: np.ndarray) -> np.ndarray:
+    """Compute the IB distortion matrix, the KL divergence between p(y|x) and q(y|xhat), in nats."""
+    # D[p(y|x) || q(y|xhat)],
+    # input shape `(x, xhat, y)`, output shape `(x, xhat)`
```

Comment:
I'd be curious to get your thoughts on using `jaxtyping` to annotate the shape of numpy arrays for methods like these: https://docs.kidger.site/jaxtyping/
It was originally written for jax, but the type annotations work with other libraries like numpy and pytorch